csrc/layernorm_kernels.cu

```diff
@@ -168,16 +177,42 @@ void rms_norm(torch::Tensor& out,     // [..., hidden_size]
   int num_tokens = input_view.numel() / hidden_size;
   int64_t input_stride = input_view.stride(-2);
 
+  // For large num_tokens, use smaller blocks to increase SM concurrency.
+  const int max_block_size = (num_tokens < 256) ? 1024 : 256;
   dim3 grid(num_tokens);
-  dim3 block(std::min(hidden_size, 1024));
   const at::cuda::OptionalCUDAGuard device_guard(device_of(input_view));
   const cudaStream_t stream = at::cuda::getCurrentCUDAStream();
   VLLM_DISPATCH_FLOATING_TYPES(
       input_view.scalar_type(), "rms_norm_kernel", [&] {
-        vllm::rms_norm_kernel<scalar_t><<<grid, block, 0, stream>>>(
-            out.data_ptr<scalar_t>(), input_view.data_ptr<scalar_t>(),
-            input_stride, weight.data_ptr<scalar_t>(), epsilon, num_tokens,
-            hidden_size);
+        const int vec_size = std::gcd(16 / sizeof(scalar_t), hidden_size);
+        const int block_size = std::min(hidden_size / vec_size, max_block_size);
+        dim3 block(block_size);
+        switch (vec_size) {
+          case 8:
```

Comment:
Can you create a dispatching macro, similar to the type dispatching macro, that does the vectorization? It can also be a regular function taking a lambda

Reply:
I have aded a dispatch macro in dispatch_utils.h. Please take a look. Thanks!